### DISCUSSION TITLE
Feature proposal: sorting keys by logical blocks

### DISCUSSION THREAD

**Discussion by andrey-helldar** [composer/composer#10073] [General]
The screenshot below shows two options for formatting the file: on the left is the format I am trying to adhere to, and on the right is the processed by [`normalize`](https://github.com/ergebnis/composer-normalize).

![image](https://user-images.githubusercontent.com/10347617/130243366-4b26c812-63ae-4874-9f31-1e9c0e478091.png)

`Normalize` is a project that sorts the keys of the `composer.json` file according to the [`composer-schema.json`](https://github.com/composer/composer/blob/master/res/composer-schema.json) schema.

I prefer to keep the "name" and "description" keys together as they contain a lot of text and separating them with a short "type" string is bad, in my opinion, as it spoils the look.

Also, I prefer to include the "config" block at the bottom of the file, as it refers to the composer rules, not the application.

I prefer to place the "support" block above the "require" block.

Thus, I try to divide all the keys into the following logical blocks:

```
- about the project
  - general information
    - name
    - description

  - project type
    - type
    - license

  - information for search
    - keywords
    - authors
    - support  

- dependencies
  - main dependencies
    - require
    - require-dev

  - autoload
    - autoload
    - autoload-dev

- composer config
  - config
  - minimum-stability
  - prefer-stable

- extended data
  - extra
```

And this project, in this case, offers the following logical scheme:

```
- about the project
  - general information
    - name

  - project type
    - type

  - general information
    - description

  - information for search
    - keywords

  - project type
    - license

  - information for search
    - authors

- dependencies
  - main dependencies
    - require
    - require-dev

- composer config
  - config

- extended data
  - extra

- dependencies
  - autoload
    - autoload
    - autoload-dev

- composer config
  - minimum-stability
  - prefer-stable

  - information for search
    - support
```

Personally, I think this formatting is illogical and makes it difficult to read the file.

Thus, I propose to consider changing the sorting of keys based on logical blocks:

```
- about the project
  - general information
    - name
    - description
    - version

  - project type
    - type
    - license

  - information
    - keywords
    - readme
    - time

  - package
    - archive
    - bin

  - contacts
    - authors
    - homepage
    - support

- dependencies
  - main dependencies
    - require
    - require-dev

  - autoload
    - autoload
    - autoload-dev

  - extended
    - abandoned
    - replace
    - conflict
    - provide
    - suggest

- composer config
  - config
  - minimum-stability
  - prefer-stable
  - target-dir
  - include-path

- scripts
  - scripts
  - scripts-descriptions

- repository
  - default-branch
  - non-feature-branches
  - repositories

- extended data
  - _comment
  - extra
  - funding
```

That is, the final schema will contain the keys in the following order:

```
- name
- description
- version
- type
- license
- keywords
- readme
- time
- archive
- bin
- authors
- homepage
- support
- require
- require-dev
- autoload
- autoload-dev
- abandoned
- replace
- conflict
- provide
- suggest
- config
- minimum-stability
- prefer-stable
- target-dir
- include-path
- scripts
- scripts-descriptions
- default-branch
- non-feature-branches
- repositories
- _comment
- extra
- funding
```

What do you think?

**Comment by Seldaek** [comment]
I agree in principle that the file is right now not sorted at all, and it's quite a mess. I wasn't aware the order was using by anyone.

I don't fully agree with your order though, and would rather suggest this:

```
- about the project
  - general information
    - name
    - description
    - license
    - type
    - abandoned (usually missing, but if present it should be prominent IMO)
    - version (usually missing, but if present it should be prominent IMO)
    - default-branch (usually missing, version related)
    - non-feature-branches (version related)

  - information
    - keywords
    - readme
    - time

- dependencies (this is one of the most important things, and shouldn't be shoved below support etc.)
  - repositories
  - _comment (main reason to have this is commenting about requirements IMO, so probably should be close to them)
  - require
  - require-dev
  - replace
  - conflict
  - provide
  - suggest
  - minimum-stability
  - prefer-stable

- autoload
  - autoload
  - autoload-dev
  - target-dir (rarely set but autoload-related)
  - include-path (rarely set but autoload-related)

- composer config
  - bin
  - archive
  - config
  - extra

- scripts
  - scripts
  - scripts-descriptions

- contacts
  - authors
  - homepage
  - support
  - funding
```

Of course we can bikeshed this to death.. but anyway if you are willing to submit a PR to reorder things (once the discussion here settles), I'd say why not do this.

**Reply by andrey-helldar** [reply]
You can get used to any sorting order, but you need to strive for the best.

I could have submitted an PR, but preferred to create a discussion first 😊

**Reply by Seldaek** [reply]
Right, although I'm sure a lot of it comes down to personal opinion, I tried to add some reasoning above. I'm happy to explain more if someone disagrees strongly.

BTW a related improvement would be to sort/group the https://getcomposer.org/doc/04-schema.md table of contents to match the order we settle on. Same for https://getcomposer.org/doc/06-config.md but that's a whole other topic ;)

**Reply by andrey-helldar** [reply]
I agree, your scheme looks more attractive to me personally, except for one block: I would transfer contacts under the block of information:

```
- about the project
  - general information
    - name
    - description
    - license
    - type
    - abandoned (usually missing, but if present it should be prominent IMO)
    - version (usually missing, but if present it should be prominent IMO)
    - default-branch (usually missing, version related)
    - non-feature-branches (version related)

  - information
    - keywords
    - readme
    - time

  - contacts
    - authors
    - homepage
    - support
    - funding

- dependencies (this is one of the most important things, and shouldn't be shoved below support etc.)
  - repositories
  - _comment (main reason to have this is commenting about requirements IMO, so probably should be close to them)
  - require
  - require-dev
  - replace
  - conflict
  - provide
  - suggest
  - minimum-stability
  - prefer-stable

- autoload
  - autoload
  - autoload-dev
  - target-dir (rarely set but autoload-related)
  - include-path (rarely set but autoload-related)

- composer config
  - bin
  - archive
  - config
  - extra

- scripts
  - scripts
  - scripts-descriptions
```

**Reply by Seldaek** [reply]
Yeah I can live with that too.. It really depends IMO sometimes I prefer it on top sometimes not, but then again that's why I do not use composer-normalize, I rather put stuff where I feel it belongs depending on the case :)

**Reply by andrey-helldar** [reply]
@Seldaek, may I offer PR or is it too early?

**Reply by Seldaek** [reply]
Sure feel free to PR against the master branch. It'll probably have to wait for 2.2.0 though, not sure I'd call this a bugfix :)

**Reply by andrey-helldar** [reply]
This is not a bug, this is a feature))

Thank you!

I will submit a PR shortly.

**Reply by andrey-helldar** [reply]
PR: https://github.com/composer/composer/pull/10089

**Reply by stof** [reply]
I would suggest keeping `repositories` below other dependency fields. This section can become quite long when using the `package` type, and is less relevant than `dependencies` when looking at open-source packages (as `repositories` is a root-only key, anything defined there for an open-source package is only relevant for the contributor setup, and so probably only for `require-dev`).

for the same reason, I would probably move `require-dev` below `conflict`, `provide` and `replace`, as `require-dev` does not affect the public metadata of the package.

**Reply by andrey-helldar** [reply]
@stof, about the `repositories` key, I agree. Initially, I suggested placing it as low as possible.

But `require-dev` personally seems to me not necessary to remove from `require`. In my opinion, they should be close, since this is, one might say, the main part for any application - the definition of dependencies for the main work and for development and testing.

**Reply by andrey-helldar** [reply]
@stof, and, to be honest, I would transfer the `funding` key to the very bottom:

![image](https://user-images.githubusercontent.com/10347617/138132102-266efc35-7e33-42f5-98ba-f3d113772e44.png)

**Reply by stof** [reply]
@andrey-helldar `conflict`, `replace` and `provide` are also part of defining the dependency problem for the main lib. that's why I suggest moving `require-dev` below them.

And `non-feature-branches` and `default-branch` are related to `version`, not to `repositories`.

Regarding `_comment`, I totally agree with @Seldaek that it should be near `require`, as explaining specific rules of the requirements is the common reason to use `_comment`

**Reply by andrey-helldar** [reply]
@stof, I understand correctly that you propose this order?

```
- name
- description
- license
- type
- abandoned
- version
- default-branch
- non-feature-branches
- keywords
- readme
- time
- authors
- homepage
- support
- funding
- _comment
- require
- replace
- conflict
- provide
- require-dev
- suggest
- minimum-stability
- prefer-stable
- autoload
- autoload-dev
- target-dir
- include-path
- bin
- archive
- config
- extra
- scripts
- scripts-descriptions
- repositories
```

I agree with the `_comment` position, but I think it is wrong to remove `require-dev` from `require`.

In my opinion, the order of the keys should be divided into logical blocks, and removing `require-dev` from `require` we will violate this order.

**Reply by stof** [reply]
I would keep repositories next to all the constraints. But after them, not before them (so after `suggest` or after `require-dev`)

> In my opinion, the order of the keys should be divided into logical blocks, and removing `require-dev` from `require` we will violate this order.

that's where we disagree about what logical blocks are. To me, there is *more* link between `require` and `conflict` (which are both about defining rules in the dependency solver for projects using the package) than between `require` and `require-dev` (as `require-dev` is root-only, and so impacts only contributors)

**Reply by andrey-helldar** [reply]
Installing dependencies, whether for production or development, is one thing - installing dependencies. And, in my opinion, they belong to the same logical block.

To understand my division, I will copy my vision of the logical blocks from the first comment:

```
- about the project
  - general information
    - name
    - description
    - version

  - project type
    - type
    - license

  - information
    - keywords
    - readme
    - time

  - package
    - archive
    - bin

  - contacts
    - authors
    - homepage
    - support

- dependencies
  - main dependencies
    - require
    - require-dev

  - autoload
    - autoload
    - autoload-dev

  - extended
    - abandoned
    - replace
    - conflict
    - provide
    - suggest

- composer config
  - config
  - minimum-stability
  - prefer-stable
  - target-dir
  - include-path

- scripts
  - scripts
  - scripts-descriptions

- repository
  - default-branch
  - non-feature-branches
  - repositories

- extended data
  - _comment
  - extra
  - funding
```

**Reply by stof** [reply]
The input of the dependency resolution is not only `require` and `require-dev`.

If you don't agree with my suggestion, I suggets that you at least take the order suggested by @Seldaek (one of the lead developers of Composer) rather than your own order.

**Reply by andrey-helldar** [reply]
Considering @Seldaek's suggestions, this order is posted above: https://github.com/composer/composer/discussions/10073#discussioncomment-1249737

But I also suggest moving the `repositories` key, at least under `prefer-stable`, rather than keeping them above the `require` block. @Seldaek, do you have any thoughts on this?

**Comment by kubawerlos** [comment]
cc. @localheinz
